### PR TITLE
Higher EMA Decay: 0.9995 for longer-memory weight averaging

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

SWA just failed catastrophically (PR #2234, p_in +55%) because uniform averaging diluted well-converged late-epoch weights with transitional snapshots. The key finding: **EMA is the superior averaging method for this model**.

Currently the model uses `--ema_decay 0.999`. This gives a half-life of ~693 steps (~2.1 epochs at ~331 steps/epoch). The EMA model effectively averages over the last ~4-5 epochs of training.

**Increasing EMA decay to 0.9995** doubles the half-life to ~1386 steps (~4.2 epochs), making the EMA average over ~8-10 epochs. This gives:
1. **Smoother averaging** over a wider region of the training trajectory
2. **More robust to late-training noise** from cosine LR annealing
3. **A middle ground between baseline EMA and SWA** — longer memory than 0.999, but still exponentially weighted (not uniform like SWA)

**Why this specific value:** 
- 0.999 → half-life ~2 epochs (current, may be too short)
- 0.9995 → half-life ~4 epochs (proposed, wider but still recent-biased)
- 0.9999 → half-life ~21 epochs (too long — would average over nearly half of training)

**Physical reasoning:** The model trains for ~145-150 epochs. In the last 10 epochs, the cosine LR approaches zero and the model makes increasingly small updates. A decay of 0.9995 averages these final fine-tuning updates more thoroughly, potentially finding a better "center" of the converged region.

**Confidence:** Medium. This is a minimal hyperparameter change with well-understood effects. The risk is that 0.9995 includes too many early-training weights (before the model has converged), diluting the average. But exponential weighting should handle this gracefully.

## Instructions

**No code changes needed.** Simply change `--ema_decay 0.999` to `--ema_decay 0.9995` in the run command.

### Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent fern --wandb_name "fern/ema-0.9995-s42" \
  --wandb_group "round19/higher-ema-decay" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.9995 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature

# Seed 73 — identical but --seed 73 --wandb_name "fern/ema-0.9995-s73"
```

Only change from baseline: `--ema_decay 0.9995` (was `0.999`)

### Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds, 2-seed avg, baseline comparison, W&B run IDs.

Also compare the EMA model and online model metrics (if reported separately). The gap between them tells us about averaging effectiveness.

## Baseline

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| p_tan  | **28.341** | < 28.34 |
| p_re   | **6.300**  | < 6.30  |

W&B: `hgml7i2r` (s42), `qic03vrg` (s73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent fern --wandb_name "fern/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### Higher EMA Decay (0.9995 vs 0.999 baseline)

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ% |
|--------|---------|---------|------------|----------|----|
| **p_in** | 12.220 | 12.277 | **12.248** | 11.979 | **+2.2%** ✗ |
| p_oodc | 7.618 | 7.610 | **7.614** | 7.643 | -0.4% ✓ |
| **p_tan** | 29.090 | 29.079 | **29.084** | 28.341 | **+2.6%** ✗ |
| **p_re** | 6.296 | 6.650 | **6.473** | 6.300 | **+2.7%** ✗ |

**val_loss (best):** s42=0.246, s73=0.246

**W&B runs:** `6wbzms95` (s42), `hm66nk7s` (s73)

**Group:** `round19/higher-ema-decay`

**Commands used:**
```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent fern --wandb_name "fern/ema-0.9995-s42" \
  --wandb_group "round19/higher-ema-decay" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.9995 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
# Seed 73 identical with --seed 73 --wandb_name "fern/ema-0.9995-s73"
```

**Training:** 149 epochs before 180-min timeout. Both seeds hit best at epoch 148.

**Peak memory:** ~55 GB (GPU 0), ~71 GB (GPU 1) — same as baseline.

### What happened

**Negative result.** Increasing EMA decay from 0.999 to 0.9995 degraded most metrics:
- **p_in +2.2%** — the longer EMA half-life appears to dilute the converged late-epoch weights with earlier, less-converged snapshots
- **p_tan +2.6%** — tandem transfer suffered the most
- **p_re +2.7%** — OOD Reynolds also regressed
- **p_oodc -0.4%** — marginal improvement, within noise

The hypothesis that wider EMA averaging would find a better "center" of the converged region did not hold. With the current cosine LR schedule (T_max=160) and ~149 effective training epochs, the model converges relatively sharply in the last ~20 epochs. A half-life of ~4.2 epochs (0.9995) includes too many under-converged snapshots from earlier training, diluting quality. The current 0.999 (half-life ~2.1 epochs) is already well-tuned for this training length.

### Suggested follow-ups
- **Lower EMA decay (0.9985 or 0.998)** could be worth exploring — if 0.999 is the sweet spot for ~150 epochs, a shorter half-life might actually sharpen the average further
- **Cosine EMA schedule** — starting with 0.999 and increasing to 0.9995 only in the final 20 epochs could capture benefits of longer averaging only when the model is fully converged

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2244 because ADVISOR: Closing — 3/4 metrics regressed (p_in +2.2%, p_tan +2.6%, p_re +2.7%). The longer half-life (4.2 epochs vs 2.1) dilutes the EMA with under-converged snapshots from earlier training. Current EMA decay 0.999 is well-tuned for our ~149-epoch training length. Good suggestion on the cosine EMA schedule (starting 0.999 → 0.9995 only in final epochs) — may revisit if we extend training length..
